### PR TITLE
Feature/1884/auto suggest parsers

### PR DESCRIPTION
**Pull request by moritz-suckow**
# Automatically suggest usable parsers based on an input file/folder/url

Issue: #1884

## Description

This adds the base functionality for automatically suggesting useful parsers to the user when starting the shell without any commands.
Now when starting the shell in default mode, the user is prompted to enter an input file.
Based on that input, all parsers are checked for usability and the user receives a list of parsers which he can all select to be executed.
After selection, the user configures each parser interactively and when done with that each configured parser is executed in parallel.

Currently, only the GitLogParser (works by looking for a .git file) and SonarImporter (works by looking for a sonar-project.properties file or an url) are supported for these suggestions.

### CODE REVIEW

File: analysis/export/CSVExporter/src/main/kotlin/de/maibornwolff/codecharta/exporter/csv/CSVExporter.kt [hunk 1]
```diff
@@ -98,6 +99,13 @@ class CSVExporter : Callable<Void>, InteractiveParser {
     }
 
     override fun getDialog(): ParserDialogInterface = ParserDialog
+    override fun isUsable(inputFile: String): Boolean {
```
Comment (ce-bo) [inline analysis/export/CSVExporter/src/main/kotlin/de/maibornwolff/codecharta/exporter/csv/CSVExporter.kt:102]: I would prefer `isApplicable` and another name for the inputFile as it is not only a file path:
```suggestion
    override fun isApplicable(resourceToBeParsed: String): Boolean {
```

File: analysis/import/GitLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/gitlogparser/GitLogParser.kt [hunk 2]
```diff
@@ -152,4 +154,23 @@ class GitLogParser(
...
+        if (inputFile.endsWith(".git")) {
+            return true
+        }
+
+        val searchFile = if (inputFile == "") {
+            File(Paths.get("").toAbsolutePath().toString())
+        } else {
+            File(inputFile)
+        }
+        val result = searchFile.walk()
```
Comment (ce-bo) [inline analysis/import/GitLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/gitlogparser/GitLogParser.kt:167]: I think we can assume to look for the .git folder in the current working directory since this is the default location where git information is located, isn't it? Thus, we could check if there is a folder in the current path named like `.git`
At the moment, all child directories are walked, aren't they?

File: analysis/import/GitLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/gitlogparser/GitLogParser.kt [hunk 3]
```diff
@@ -152,4 +154,23 @@ class GitLogParser(
...
+                .asSequence()
+                .map { it.name }
+                .filter { it.endsWith(".git") }
```
Comment (ce-bo) [inline analysis/import/GitLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/gitlogparser/GitLogParser.kt:170]: i think you can directly return by using `any()` instead of `filter()`.

File: analysis/import/GitLogParser/src/test/kotlin/de/maibornwolff/codecharta/importer/gitlogparser/GitLogParserTest.kt [hunk 4]
```diff
@@ -62,4 +82,18 @@ class GitLogParserTest {
 
         assertTrue(file.exists())
     }
+
+    @ParameterizedTest
+    @MethodSource("provideValidInputFiles")
+    fun `should return true with valid git repository as input for isUsable`(inputFile: String) {
```
Comment (ce-bo) [inline analysis/import/GitLogParser/src/test/kotlin/de/maibornwolff/codecharta/importer/gitlogparser/GitLogParserTest.kt:88]: I would not call the method name `isUsable` in the test name because this causes a depencency/connascence (https://www.maibornwolff.de/know-how/connascence-regeln-fuer-gutes-software-design/)

Lets name the test from a business perspective and omit the technical "HOW":
```suggestion
    fun `should be identified as applicable for given directory path containing a git folder`(inputFile: String) {
```

more candidates would be `should NOT be identified as applicable if no git folder is present at given path`

File: analysis/import/SonarImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/sonar/SonarImporterMain.kt [hunk 5]
```diff
@@ -93,4 +95,26 @@ class SonarImporterMain(
...
 
     override fun getDialog(): ParserDialogInterface = ParserDialog
+    override fun isUsable(inputFile: String): Boolean {
+        val trimmedInput = inputFile.trim()
+        // TODO: Is this sufficient to check for urls?
+        if (trimmedInput.startsWith("http://") || trimmedInput.startsWith("https://")) {
+            return true
+        }
+
+        if (trimmedInput.endsWith("sonar-project.properties")) {
```
Comment (ce-bo) [inline analysis/import/SonarImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/sonar/SonarImporterMain.kt:105]: Just curious... Why using `endsWith?`

File: analysis/import/SonarImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/sonar/SonarImporterMain.kt [hunk 6]
```diff
@@ -93,4 +95,26 @@ class SonarImporterMain(
...
+            return true
+        }
+
+        val searchFile = File(inputFile)
+        val files = searchFile.walk().asSequence()
+                .filter { it.isFile }
+                .map { it.name }
```
Comment (ce-bo) [inline analysis/import/SonarImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/sonar/SonarImporterMain.kt:112]: We could shorten this to 
```suggestion
return trimmedInput.startsWith("http://") || trimmedInput.startsWith("https://") || trimmedInput.endsWith("sonar-project.properties") || searchFile.walk().asSequence()
                .filter { it.isFile }
                .any{ it.name == "sonar-project.properties" }
```

File: analysis/tools/InteractiveParser/src/main/kotlin/de/maibornwolff/codecharta/tools/interactiveparser/util/InteractiveParserHelper.kt [hunk 7]
```diff
@@ -0,0 +1,84 @@
+package de.maibornwolff.codecharta.tools.interactiveparser.util
```
Comment (ce-bo) [inline analysis/tools/InteractiveParser/src/main/kotlin/de/maibornwolff/codecharta/tools/interactiveparser/util/InteractiveParserHelper.kt:1]: Thanks for the refactoring!

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt [hunk 8]
```diff
@@ -75,15 +87,80 @@ class Ccsh : Callable<Void?> {
...
+                val configuredParsers = offerInteractiveParserSuggestions(commandLine)
+                if (configuredParsers.isEmpty()) {
+                    return 0
+                }
+
+                return executeConfiguredParsers(commandLine, configuredParsers)
+            } else if (isParserUnknown(args, commandLine) || args.contains("--interactive") || args.contains("-i")) {
+                executeInteractiveParser(commandLine)
             } else {
-                return commandLine.execute(*sanitizeArgs(args))
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:81]: There is no value returned anymore in this case.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt [hunk 9]
```diff
@@ -75,15 +87,80 @@ class Ccsh : Callable<Void?> {
...
+                commandLine.execute(*sanitizeArgs(args))
             }
         }
 
+        fun offerInteractiveParserSuggestions(commandLine: CommandLine): Map<String, List<String>> {
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:104]: This method is more like `get...` method.
You could also think about to extract this functionality to InteractiveParserSuggestionDialog class.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt [hunk 10]
```diff
@@ -75,15 +87,80 @@ class Ccsh : Callable<Void?> {
...
+            val inputFile: String = KInquirer.promptInput(
+                    message = "What is the name of the input file/folder/url?",
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:106]: I would Maybe do it like this:
Which path should be scanned?
Hint: You can provide a directory path / file path / sonar url

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt [hunk 11]
```diff
@@ -75,15 +87,80 @@ class Ccsh : Callable<Void?> {
...
+            val selectedParsers = ParserService.offerParserSuggestions(commandLine, PicocliParserRepository(), inputFile)
+
+            if (selectedParsers.isEmpty()) {
+                println(NO_USABLE_PARSER_FOUND_MESSAGE)
+                return emptyMap()
+            }
+
+            val shouldRunConfiguredParsers: Boolean =
+                    KInquirer.promptConfirm(
+                            message = "Do you want to run all configured parsers?",
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:118]:
```suggestion
                            message = "Do you want to run all configured parsers now?",
```

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt [hunk 12]
```diff
@@ -75,15 +87,80 @@ class Ccsh : Callable<Void?> {
...
+                            default = true)
+
+            if (!shouldRunConfiguredParsers) {
+                return emptyMap()
+            }
+
+            return (ParserService.configureParserSelection(commandLine, PicocliParserRepository(), selectedParsers))
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:125]:
```suggestion
            return !shouldRunConfiguredParsers ? emptyMap() : ParserService.configureParserSelection(commandLine, PicocliParserRepository(), selectedParsers)
```

File: analysis/import/GitLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/gitlogparser/GitLogParser.kt (hunk 2 above)
Comment (moritz-suckow) [reply analysis/import/GitLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/gitlogparser/GitLogParser.kt:167]: Yes currently we iterate through all child directories and look for a .git file and yes I think you are right, we should probably just look at the given/current directory for a .git file, since we specifically ask for that folder anyway.

File: analysis/import/SonarImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/sonar/SonarImporterMain.kt (hunk 5 above)
Comment (moritz-suckow) [reply analysis/import/SonarImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/sonar/SonarImporterMain.kt:105]: Well if a project contains this file it is very probable that it is a sonar analyzed project, so suggesting sonar made sense to me.

About your second comment: I just copy pasted it from the GitLogParser and forgot to change the method

File: analysis/import/SonarImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/sonar/SonarImporterMain.kt (hunk 6 above)
Comment (moritz-suckow) [reply analysis/import/SonarImporter/src/main/kotlin/de/maibornwolff/codecharta/importer/sonar/SonarImporterMain.kt:112]: Sure we could do that - although I find it a bit ugly to look at :D

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt [hunk 13]
```diff
@@ -75,15 +87,80 @@ class Ccsh : Callable<Void?> {
...
+        }
+
+        fun executeConfiguredParsers(commandLine: CommandLine, configuredParsers: Map<String, List<String>>): Int {
+            val exitCode = AtomicInteger(0)
+            val threadPool = Executors.newFixedThreadPool(configuredParsers.size)
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:130]: Is there a way to calculate an upper bound of threads to be started?

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt (hunk 8 above)
Comment (moritz-suckow) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:81]: I moved the return in front of the if - should this not just return the old value or did I understand this wrong?

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt [hunk 14]
```diff
@@ -75,15 +87,80 @@ class Ccsh : Callable<Void?> {
...
+                    if (currentExitCode != 0) {
+                        println("Code: $currentExitCode")
+                        exitCode.set(currentExitCode)
+                    }
+                }
+            }
+            threadPool.shutdown()
+            threadPool.awaitTermination(1, TimeUnit.DAYS)
+
+            println("Code: ${exitCode.get()}")
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:143]: Please don't mix up println and logger outputs.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt (hunk 10 above)
Comment (moritz-suckow) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:106]: Sure that sounds better!

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt [hunk 15]
```diff
@@ -75,15 +87,80 @@ class Ccsh : Callable<Void?> {
...
+            if (exitCode.get() != 0) {
+                return exitCode.get()
+            }
+            // Improvement: Try to extract merge commands before so user does not have to configure merge args?
+            return ParserService.executeSelectedParser(commandLine, "merge")
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:148]: Is this call stable? Otherwise let's notify the users in a first step that certain .cc.json files have been created and that they can visualize or merge it by themselves.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt [hunk 16]
```diff
@@ -1,55 +1,83 @@
...
-            return extractParserName(selectedParser)
+            return parserRepository.extractParserName(selectedParser)
         }
 
         fun executeSelectedParser(commandLine: CommandLine, selectedParser: String): Int {
             val subCommand = commandLine.subcommands.getValue(selectedParser)
             val parserObject = subCommand.commandSpec.userObject()
             val interactive = parserObject as? InteractiveParser
-            if (interactive != null) {
+            return if (interactive != null) {
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt:56]: The function has Int as return type. Why did you remove the return statement? :-)

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt (hunk 13 above)
Comment (moritz-suckow) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:130]: Good idea, for now I decided against putting more effort into this because the number of unique parsers is so low.
If we implemented all parsers to be suggestable, we would have 15 or something parsers. Even if we doubled that amount somehow - is starting 30 threads too much?

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/repository/IParserRepository.kt [hunk 17]
```diff
@@ -0,0 +1,11 @@
+package de.maibornwolff.codecharta.tools.ccsh.parser.repository
+
+import de.maibornwolff.codecharta.tools.interactiveparser.InteractiveParser
+interface IParserRepository<T> {
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/repository/IParserRepository.kt:4]: As far as I know, we do not prefix our interfaces with `I`. As such, I would stick to the current approach and remove it.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/repository/PicocliParserRepository.kt [hunk 18]
```diff
@@ -0,0 +1,64 @@
+package de.maibornwolff.codecharta.tools.ccsh.parser.repository
+
+import de.maibornwolff.codecharta.tools.interactiveparser.InteractiveParser
+import picocli.CommandLine
+
+class PicocliParserRepository : IParserRepository<CommandLine> {
+
+    override fun getParserNames(dataSource: CommandLine): List<String> {
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/repository/PicocliParserRepository.kt:8]: This method only returns interactive parsers but the method does not show this.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/repository/PicocliParserRepository.kt [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+        val subCommands = dataSource.subcommands.values
+        return subCommands.mapNotNull { subCommand ->
+            val parserName = subCommand.commandName
+            if (subCommand.commandSpec.userObject() is InteractiveParser) {
+                parserName
+            } else null
+        }
+    }
+
+    override fun getParserNamesWithDescription(dataSource: CommandLine): List<String> {
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/repository/PicocliParserRepository.kt:18]: Same here...

```suggestion
    override fun getInteractiveParserDescriptions(dataSource: CommandLine): List<String> {
```

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/repository/PicocliParserRepository.kt [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+            if (parser.isUsable(inputFile)) {
+                usableParsers.add(parser.getName())
+            }
+        }
+        return usableParsers
+    }
+
+    override fun getParser(dataSource: CommandLine, name: String): InteractiveParser? {
+        val subCommand = dataSource.subcommands.getValue(name)
+        val parserObject = subCommand.commandSpec.userObject()
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/repository/PicocliParserRepository.kt:60]: Please make sure to not cause Nullpointer exceptions if name is not found.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt (hunk 16 above)
Comment (moritz-suckow) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/parser/ParserService.kt:56]: > I moved the return in front of the if - should this not just return the old value or did I understand this wrong?
Same reasoning,  I just moved the return in front of the if - I think I will revert that change though, since it is a bit confusing.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt (hunk 15 above)
Comment (moritz-suckow) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:148]: I am not sure if this is stable since I only did some basic testing with the merger, and in those tests this solution worked.

We could notify the user as you said and add that functionality with a bit more testing in a new issue/pr if that is okay with you?

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt (hunk 9 above)
Comment (moritz-suckow) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:104]: I refactored it a bit: Now it only returns all applicable parsers but does not use the Kinquirer prompt. This is done in the calling function (ccsh.kt) now since all other prompts are also created there.

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt (hunk 13 above)
Comment (moritz-suckow) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:130]: As we decided we will limit the amount of threads that we start by the following formula:
min(parsersToBeStarted, CPU-Cores)

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt (hunk 15 above)
Comment (moritz-suckow) [reply analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:148]: For now I will just print a message informing the user about the result and that he can manually use the merge-tool to combine all parser results

File: analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt [hunk 21]
```diff
@@ -87,73 +89,85 @@ class Ccsh : Callable<Void?> {
...
-            return (ParserService.configureParserSelection(commandLine, PicocliParserRepository(), selectedParsers))
+            return if (shouldRunConfiguredParsers) {
+                ParserService.configureParserSelection(commandLine, PicocliParserRepository(), selectedParsers)
+            } else { emptyMap() }
         }
 
         fun executeConfiguredParsers(commandLine: CommandLine, configuredParsers: Map<String, List<String>>): Int {
             val exitCode = AtomicInteger(0)
-            val threadPool = Executors.newFixedThreadPool(configuredParsers.size)
+            val numberOfThreadsToBeStarted = min(configuredParsers.size, Runtime.getRuntime().availableProcessors())
```
Comment (ce-bo) [inline analysis/tools/ccsh/src/main/kotlin/de/maibornwolff/codecharta/tools/ccsh/Ccsh.kt:116]: I like verbose variable names :-)

**Comment by ce-bo** [comment]
Please add a changelog entry.

**Review by ce-bo — APPROVED** [review]

**Comment by moritz-suckow** [comment]
FeaturePreviewAutoSuggestion
